src/pcap/agent.go

```diff
@@ -0,0 +1,367 @@
+package pcap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/pcap"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+// Agent is the central struct to which the handlers are attached.
+type Agent struct {
+	// done is used to gracefully shut down the agent, all ongoing streams terminate
+	// whenever this channel is closed.
+	done chan struct{}
+	// wg tracks any running streams.
+	// TODO: expose as metric?
+	wg sync.WaitGroup
+	// log carries the logger all session loggers are derived from.
+	log     *zap.Logger
+	bufConf BufferConf
+
+	UnimplementedAgentServer
+}
+
+// NewAgent creates a new ready-to-use agent. If the given logger is nil zap.L will
+// be used.
+func NewAgent(log *zap.Logger, bufConf BufferConf) (*Agent, error) {
+	if log == nil {
+		log = zap.L()
+	}
+	return &Agent{
+		done:    make(chan struct{}),
+		log:     log,
+		bufConf: bufConf,
+	}, nil
+}
+
+// Stop the server. This will gracefully stop any captures that are currently running
+// by closing Agent.done. Further calls to Stop have no effect.
+func (a *Agent) Stop() {
+	select {
+	case <-a.done:
+		// if the channel is already closed, we do nothing
+	default:
+		// otherwise the channel is still open and we close it
+		close(a.done)
+	}
+}
+
+// Wait for all open streams to terminate.
+func (a *Agent) Wait() {
+	a.wg.Wait()
+}
+
+// draining returns true after Agent.Stop has been called.
+func (a *Agent) draining() bool {
+	select {
+	case <-a.done:
+		// we only get here if the channel is closed since it is never written to
+		return true
+	default:
+		// channel is still open
+		return false
+	}
+}
+
+// Status handler for the pcap-agent. See AgentServer.Status documentation for details.
+func (a *Agent) Status(_ context.Context, _ *StatusRequest) (*StatusResponse, error) {
+	s := &StatusResponse{
+		CompatibilityLevel: CompatibilityLevel,
+		Health:             Health_UP,
+		Status:             "ok",
+	}
+
+	if a.draining() {
+		s.Health = Health_DRAINING
+		s.Status = "agent is shutting down"
```

Comment:
hmm this seems to conflate status and statusMessage. the agent's status should be both DRAINING and SHUTTING DOWN or STOPPING and there should be a map of status codes to messages that are returned to clients. if we say DRAINING always implies "shutting down" there may only be one status field internally, but the message should not be hard-coded here in any case.

Reply:
The idea behind the two fields was that the enum should be used for actionable things like draining that the API/client needs to consider and the message is something that can be given to the user as a reason. In this case the agent is draining but the API doesn't care why so the status is vague but the user would probably like to know why his capture didn't start as expected -> because the agent is currently being stopped.

Reply:
Maybe the fact that we include a draining state in the health options is a bad idea since it does conflate two different kinds of status (like you mentioned).

Realistically the API/client only cares whether the agent is UP or DOWN. @b1tamara and my proposal would be to only have those two statuses and put any further details in the free-text field.

Reply:
During an offline discussion we agreed to remove the enum `Health` and replace it with a simple `healthy` boolean in the status response because this field should be limited to whether the agent is available for capture or not.

Reply:
as discussed we will change health to status and status to message, no other operating states for now